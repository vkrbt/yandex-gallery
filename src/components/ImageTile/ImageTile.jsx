import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { Img } from '../Img/Img';
import { calculateRatio, calculateWidth } from '../../helpers/image';

export class ImageTile extends PureComponent {
  constructor(props) {
    super(props);

    const imageRatio = calculateRatio(props.image.width, props.image.height);

    this.state = {
      isError: false,
      width: calculateWidth(props.tileHeight, imageRatio),
    };

    this.handleImageError = this.handleImageError.bind(this);
    this.handleSelect = this.handleSelect.bind(this);
  }

  handleImageError() {
    this.setState({
      isError: true,
    });
  }

  handleSelect() {
    this.props.handleSelect(this.props.image);
  }

  render() {
    const { isError } = this.state;
    return !isError ? (
      <div
        role="button"
        tabIndex="0"
        className="image-tile"
        style={{ backgroundColor: this.props.image.color, flexBasis: this.state.width }}
        onClick={this.handleSelect}
        onKeyPress={this.handleSelect}
      >
        <Img className="image-tile__image" src={this.props.image.urls.small} onImageError={this.handleImageError} />
        <div />
      </div>
    ) : null;
  }
}

ImageTile.defaultProps = {
  tileHeight: 200,
};

ImageTile.propTypes = {
  image: PropTypes.shape({
    id: PropTypes.string.isRequired,
    width: PropTypes.number.isRequired,
    height: PropTypes.number.isRequired,
    color: PropTypes.string.isRequired,
    urls: PropTypes.shape({
      small: PropTypes.string.isRequired,
      regular: PropTypes.string.isRequired,
    }).isRequired,
  }).isRequired,
  handleSelect: PropTypes.func.isRequired,
  tileHeight: PropTypes.number,
};
