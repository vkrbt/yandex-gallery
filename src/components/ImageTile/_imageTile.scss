$hover-animation: 0.1s ease-out;

.image-tile {
  position: relative;

  overflow: hidden;
  flex-grow: 1;

  height: 200px;
  margin: $tile-padding;
  margin-bottom: 0;
  margin-left: 0;

  cursor: pointer;

  outline: none;
  &::after {
    position: absolute;
    top: 0;
    right: 0;
    bottom: 0;
    left: 0;

    content: '';
    transition: $hover-animation;

    opacity: 0;
    background-color: #000;
  }
  &:hover,
  &:focus {
    @media not all and (pointer: coarse) {
      &::after {
        visibility: visible;

        opacity: 0.3;
      }
    }
  }
  &__image {
    flex-grow: 1;

    min-width: 100%;
    max-width: 100%;
    height: inherit;

    object-fit: cover;
  }
  &__description {
    font-size: 0.9rem;
    line-height: 1.1rem;
    padding: 5px;
    color: #fff;
    width: 100%;
    position: absolute;
    bottom: 0;
    background-color: rgba(0, 0, 0, 0.5);
    z-index: 1;
    opacity: 0;
    transform: translate3d(0, 100%, 0);
    transform-origin: 100% 100%;
    transition: $hover-animation;
  }
  &:hover,
  &:focus {
    .image-tile__description {
      opacity: 1;
      transform: translate3d(0, 0, 0);
    }
  }
  @media (pointer: coarse) {
    .image-tile__description {
      opacity: 1;
      transform: translate3d(0, 0, 0);
    }
  }
  &__link {
    color: #bbb;
    &:hover {
      text-decoration: none;
    }
  }
}
