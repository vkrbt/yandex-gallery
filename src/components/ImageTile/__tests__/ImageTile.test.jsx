import React from 'react';
import { shallow } from 'enzyme';
import { ImageTile } from '../ImageTile';

const image = {
  id: 'id',
  width: 500,
  height: 800,
  color: '#FFF',
  urls: {
    small: 'some url',
    regular: 'some url',
  },
  user: {
    name: 'name',
    username: 'username',
  },
};
const props = {
  image,
  handleSelect: jest.fn(),
};

describe('ImageTile', () => {
  it('should being rendered normally', () => {
    const wrapper = shallow(<ImageTile {...props} />);
    expect(wrapper).toMatchSnapshot();
  });
});
