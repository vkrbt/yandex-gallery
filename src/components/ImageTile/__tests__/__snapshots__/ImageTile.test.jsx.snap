// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageTile should being rendered normally 1`] = `
<div
  className="image-tile"
  onClick={[Function]}
  onKeyPress={[Function]}
  role="button"
  style={
    Object {
      "backgroundColor": "#FFF",
      "flexBasis": "125px",
    }
  }
  tabIndex="0"
>
  <Img
    alt=""
    className="image-tile__image"
    src="some url"
  />
  <p
    className="image-tile__description"
  >
    Photo by
     
    <a
      className="image-tile__link"
      href="https://unsplash.com/@username?utm_source=yandex-gallery&utm_medium=referral"
      onClick={[Function]}
    >
      name
    </a>
     
    on
     
    <a
      className="image-tile__link"
      href="https://unsplash.com/?utm_source=yandex-gallery&utm_medium=referral"
      onClick={[Function]}
    >
      Unsplash
    </a>
  </p>
</div>
`;
